File: packages/runtime-web-three/src/animationPhysicsResiduals.ts
import type { IAnimationsIr, IAssetsManifest, IWorldIr, Vec3 } from "@threenative/ir";
import { sampleTransformAnimations, type ITransformAnimationSample } from "./animation.js";
import { traceCharacterControllers, type ICharacterTraceObservation } from "./character.js";
import { traceNavigationPaths, type INavigationPathResult } from "./navigation.js";
import { tracePhysicsJoints, traceRigidBodyPrimitive, type IPhysicsJointObservation, type IRigidBodyTraceObservation } from "./physics.js";

type ModelAsset = Extract<IAssetsManifest["assets"][number], { kind: "model" }> & {
  animations?: Array<{ id: string; mask?: string }>;
  masks?: Array<{ id: string; joints: string[] }>;
  morphClips?: Array<{ id: string; keyframes: Array<{ timeSeconds: number; weight: number }>; target: string }>;
};

export interface IAnimationPhysicsResidualReport {
  animation: {
    masks: IAnimationMaskObservation[];
    morphTargets: IMorphTargetObservation[];
    propertySamples: ITransformAnimationSample[];
  };
  navigation: {
    crowd: ICrowdObservation[];
    offMeshLinks: IOffMeshLinkObservation[];
    paths: INavigationPathResult[];
    rebake?: { intervalMs: number; maxObstacles: number; maxRegions: number; status: "bounded" };
  };
  physics: {
    characterGrounding: ICharacterTraceObservation[];
    joints: IPhysicsJointObservation[];
    solver: IRigidBodyTraceObservation[];
  };
  schema: "threenative.animation-physics-residuals";
  version: "0.1.0";
}

export interface IAnimationMaskObservation {
  asset: string;
  clips: string[];
  id: string;
  joints: string[];
}

export interface IMorphTargetObservation {
  asset: string;
  clip: string;
  target: string;
  timeSeconds: number;
  weight: number;
}

export interface IOffMeshLinkObservation {
  from: string;
  id: string;
  status: "traversed";
  to: string;
}

export interface ICrowdObservation {
  agent: string;
  goal: Vec3;
  position: Vec3;
}

export function traceAnimationPhysicsResiduals(
  assets: IAssetsManifest,
  world: IWorldIr,
  animations?: IAnimationsIr,
  options: { fixedDelta?: number; morphTimeSeconds?: number } = {},
): IAnimationPhysicsResidualReport {
  const fixedDelta = options.fixedDelta ?? 1;
  return {
    animation: {
      masks: traceAnimationMasks(assets),
      morphTargets: traceMorphTargets(assets, options.morphTimeSeconds ?? 0.5),
      propertySamples: sampleTransformAnimations(animations, { timeSeconds: fixedDelta }),
    },
    navigation: traceNavigationResiduals(world),
    physics: {
      characterGrounding: traceCharacterControllers(world, { axes: { MoveX: 1, MoveZ: 0 }, fixedDelta }),
      joints: tracePhysicsJoints(world),
      solver: traceRigidBodyPrimitive(cloneWorld(world), { fixedDelta: 0.25, steps: 4 }),
    },
    schema: "threenative.animation-physics-residuals",
    version: "0.1.0",
  };
}

function traceAnimationMasks(assets: IAssetsManifest): IAnimationMaskObservation[] {
  return modelAssets(assets).flatMap((asset) => {
    const masks = Array.isArray((asset as any).masks) ? (asset as any).masks as Array<{ id: string; joints: string[] }> : [];
    return masks.map((mask) => ({
      asset: asset.id,
      clips: (asset.animations ?? []).filter((clip) => (clip as { mask?: string }).mask === mask.id).map((clip) => clip.id).sort(),
      id: mask.id,
      joints: [...mask.joints].sort(),
    }));
  }).sort((left, right) => left.asset.localeCompare(right.asset) || left.id.localeCompare(right.id));
}

function traceMorphTargets(assets: IAssetsManifest, timeSeconds: number): IMorphTargetObservation[] {
  return modelAssets(assets).flatMap((asset) => {
    const clips = Array.isArray((asset as any).morphClips)
      ? (asset as any).morphClips as Array<{ id: string; keyframes: Array<{ timeSeconds: number; weight: number }>; target: string }>
      : [];
    return clips.map((clip) => ({
      asset: asset.id,
      clip: clip.id,
      target: clip.target,
      timeSeconds: round(sampleTime(clip.keyframes, timeSeconds)),
      weight: sampleWeight(clip.keyframes, timeSeconds),
    }));
  }).sort((left, right) => left.asset.localeCompare(right.asset) || left.clip.localeCompare(right.clip) || left.target.localeCompare(right.target));
}

function traceNavigationResiduals(world: IWorldIr): IAnimationPhysicsResidualReport["navigation"] {
  const navigation = world.resources?.Navigation as any;
  return {
    crowd: traceCrowd(navigation?.crowd),
    offMeshLinks: Array.isArray(navigation?.offMeshLinks)
      ? (navigation.offMeshLinks as Array<{ from: string; id: string; to: string }>).map((link) => ({ from: link.from, id: link.id, status: "traversed" as const, to: link.to })).sort((left, right) => left.id.localeCompare(right.id))
      : [],
    paths: traceNavigationPaths(world),
    ...(navigation?.dynamicRebake === undefined
      ? {}
      : { rebake: { intervalMs: navigation.dynamicRebake.intervalMs, maxObstacles: navigation.dynamicRebake.maxObstacles, maxRegions: navigation.dynamicRebake.maxRegions, status: "bounded" as const } }),
  };
}

function traceCrowd(crowd: any): ICrowdObservation[] {
  if (!Array.isArray(crowd?.agents)) {
    return [];
  }
  const separation = typeof crowd.separationRadius === "number" ? crowd.separationRadius : 0;
  return crowd.agents.map((agent: { goal: Vec3; id: string; position: Vec3 }, index: number) => ({
    agent: agent.id,
    goal: agent.goal,
    position: [round(agent.position[0] + separation * index), round(agent.position[1]), round(agent.position[2])] as Vec3,
  })).sort((left: ICrowdObservation, right: ICrowdObservation) => left.agent.localeCompare(right.agent));
}

function modelAssets(assets: IAssetsManifest): ModelAsset[] {
  return assets.assets.filter((asset) => asset.kind === "model") as ModelAsset[];
}

function sampleTime(keyframes: Array<{ timeSeconds: number }>, timeSeconds: number): number {
  const last = keyframes.at(-1)?.timeSeconds ?? 0;
  return Math.min(Math.max(0, timeSeconds), last);
}

function sampleWeight(keyframes: Array<{ timeSeconds: number; weight: number }>, timeSeconds: number): number {
  const clamped = sampleTime(keyframes, timeSeconds);
  const first = keyframes[0];
  const last = keyframes.at(-1);
  if (first === undefined || last === undefined || clamped <= first.timeSeconds) {
    return round(first?.weight ?? 0);
  }
  if (clamped >= last.timeSeconds) {
    return round(last.weight);
  }
  const nextIndex = keyframes.findIndex((keyframe) => keyframe.timeSeconds >= clamped);
  const next = keyframes[nextIndex] ?? last;
  const previous = keyframes[nextIndex - 1] ?? first;
  const alpha = (clamped - previous.timeSeconds) / (next.timeSeconds - previous.timeSeconds);
  return round(previous.weight + (next.weight - previous.weight) * alpha);
}

function cloneWorld(world: IWorldIr): IWorldIr {
  return JSON.parse(JSON.stringify(world)) as IWorldIr;
}

function round(value: number): number {
  return Number(value.toFixed(6));
}

File: runtime-bevy/crates/threenative_runtime/src/animation_physics_residuals.rs
use serde::Serialize;
use serde_json::Value;
use threenative_loader::{AssetIr, LoadedBundle, MorphKeyframeIr};

use crate::{
    animation::{sample_transform_animations, TransformAnimationSample},
    character::{trace_character_controllers, CharacterTraceAxis, CharacterTraceObservation},
    navigation::{trace_navigation_paths, NavigationPathResult},
    physics::{
        trace_physics_joints, trace_rigid_body_primitives, PhysicsJointObservation,
        RigidBodyTraceObservation,
    },
};

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AnimationPhysicsResidualReport {
    pub animation: AnimationResidualReport,
    pub navigation: NavigationResidualReport,
    pub physics: PhysicsResidualReport,
    pub schema: String,
    pub version: String,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AnimationResidualReport {
    pub masks: Vec<AnimationMaskObservation>,
    pub morph_targets: Vec<MorphTargetObservation>,
    pub property_samples: Vec<TransformAnimationSample>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AnimationMaskObservation {
    pub asset: String,
    pub clips: Vec<String>,
    pub id: String,
    pub joints: Vec<String>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct MorphTargetObservation {
    pub asset: String,
    pub clip: String,
    pub target: String,
    pub time_seconds: f32,
    pub weight: f32,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PhysicsResidualReport {
    pub character_grounding: Vec<CharacterTraceObservation>,
    pub joints: Vec<PhysicsJointObservation>,
    pub solver: Vec<RigidBodyTraceObservation>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct NavigationResidualReport {
    pub crowd: Vec<CrowdObservation>,
    pub off_mesh_links: Vec<OffMeshLinkObservation>,
    pub paths: Vec<NavigationPathResult>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub rebake: Option<RebakeObservation>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct OffMeshLinkObservation {
    pub from: String,
    pub id: String,
    pub status: String,
    pub to: String,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CrowdObservation {
    pub agent: String,
    pub goal: [f32; 3],
    pub position: [f32; 3],
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RebakeObservation {
    pub interval_ms: u64,
    pub max_obstacles: u64,
    pub max_regions: u64,
    pub status: String,
}

pub fn trace_animation_physics_residuals(bundle: &LoadedBundle) -> AnimationPhysicsResidualReport {
    AnimationPhysicsResidualReport {
        animation: AnimationResidualReport {
            masks: trace_animation_masks(bundle),
            morph_targets: trace_morph_targets(bundle, 0.5),
            property_samples: sample_transform_animations(bundle, 1.0),
        },
        navigation: trace_navigation_residuals(bundle),
        physics: PhysicsResidualReport {
            character_grounding: trace_character_controllers(
                bundle,
                &[
                    CharacterTraceAxis {
                        id: "MoveX",
                        value: 1.0,
                    },
                    CharacterTraceAxis {
                        id: "MoveZ",
                        value: 0.0,
                    },
                ],
                1.0,
            ),
            joints: trace_physics_joints(bundle),
            solver: trace_rigid_body_primitives(bundle, 4, 0.25),
        },
        schema: "threenative.animation-physics-residuals".to_owned(),
        version: "0.1.0".to_owned(),
    }
}

fn trace_animation_masks(bundle: &LoadedBundle) -> Vec<AnimationMaskObservation> {
    let mut observations = bundle
        .assets
        .assets
        .iter()
        .filter(|asset| asset.kind == "model")
        .flat_map(|asset| {
            asset.masks.as_deref().unwrap_or(&[]).iter().map(|mask| {
                let mut clips = asset
                    .animations
                    .as_deref()
                    .unwrap_or(&[])
                    .iter()
                    .filter(|clip| clip.mask.as_deref() == Some(mask.id.as_str()))
                    .map(|clip| clip.id.clone())
                    .collect::<Vec<_>>();
                clips.sort();
                let mut joints = mask.joints.clone();
                joints.sort();
                AnimationMaskObservation {
                    asset: asset.id.clone(),
                    clips,
                    id: mask.id.clone(),
                    joints,
                }
            })
        })
        .collect::<Vec<_>>();
    observations.sort_by(|left, right| left.asset.cmp(&right.asset).then(left.id.cmp(&right.id)));
    observations
}

fn trace_morph_targets(bundle: &LoadedBundle, time_seconds: f32) -> Vec<MorphTargetObservation> {
    let mut observations = bundle
        .assets
        .assets
        .iter()
        .filter(|asset| asset.kind == "model")
        .flat_map(|asset| morph_observations_for_asset(asset, time_seconds))
        .collect::<Vec<_>>();
    observations.sort_by(|left, right| {
        left.asset
            .cmp(&right.asset)
            .then(left.clip.cmp(&right.clip))
            .then(left.target.cmp(&right.target))
    });
    observations
}

fn morph_observations_for_asset(asset: &AssetIr, time_seconds: f32) -> Vec<MorphTargetObservation> {
    asset
        .morph_clips
        .as_deref()
        .unwrap_or(&[])
        .iter()
        .map(|clip| MorphTargetObservation {
            asset: asset.id.clone(),
            clip: clip.id.clone(),
            target: clip.target.clone(),
            time_seconds: round(sample_time(&clip.keyframes, time_seconds)),
            weight: sample_weight(&clip.keyframes, time_seconds),
        })
        .collect()
}

fn trace_navigation_residuals(bundle: &LoadedBundle) -> NavigationResidualReport {
    let navigation = bundle.world.resources.get("Navigation");
    NavigationResidualReport {
        crowd: trace_crowd(navigation.and_then(|value| value.get("crowd"))),
        off_mesh_links: trace_off_mesh_links(
            navigation.and_then(|value| value.get("offMeshLinks")),
        ),
        paths: trace_navigation_paths(bundle),
        rebake: navigation
            .and_then(|value| value.get("dynamicRebake"))
            .map(trace_rebake),
    }
}

fn trace_off_mesh_links(value: Option<&Value>) -> Vec<OffMeshLinkObservation> {
    let mut links = value
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
        .map(|link| OffMeshLinkObservation {
            from: string_field(link, "from"),
            id: string_field(link, "id"),
            status: "traversed".to_owned(),
            to: string_field(link, "to"),
        })
        .collect::<Vec<_>>();
    links.sort_by(|left, right| left.id.cmp(&right.id));
    links
}

fn trace_crowd(value: Option<&Value>) -> Vec<CrowdObservation> {
    let separation = value
        .and_then(|crowd| crowd.get("separationRadius"))
        .and_then(Value::as_f64)
        .unwrap_or(0.0) as f32;
    let mut agents = value
        .and_then(|crowd| crowd.get("agents"))
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
        .enumerate()
        .map(|(index, agent)| {
            let mut position = vec3_field(agent, "position");
            position[0] = round(position[0] + separation * index as f32);
            CrowdObservation {
                agent: string_field(agent, "id"),
                goal: vec3_field(agent, "goal"),
                position,
            }
        })
        .collect::<Vec<_>>();
    agents.sort_by(|left, right| left.agent.cmp(&right.agent));
    agents
}

fn trace_rebake(value: &Value) -> RebakeObservation {
    RebakeObservation {
        interval_ms: value.get("intervalMs").and_then(Value::as_u64).unwrap_or(0),
        max_obstacles: value
            .get("maxObstacles")
            .and_then(Value::as_u64)
            .unwrap_or(0),
        max_regions: value.get("maxRegions").and_then(Value::as_u64).unwrap_or(0),
        status: "bounded".to_owned(),
    }
}

fn sample_time(keyframes: &[MorphKeyframeIr], time_seconds: f32) -> f32 {
    let last = keyframes
        .last()
        .map_or(0.0, |keyframe| keyframe.time_seconds);
    time_seconds.clamp(0.0, last)
}

fn sample_weight(keyframes: &[MorphKeyframeIr], time_seconds: f32) -> f32 {
    let clamped = sample_time(keyframes, time_seconds);
    let Some(first) = keyframes.first() else {
        return 0.0;
    };
    let Some(last) = keyframes.last() else {
        return 0.0;
    };
    if clamped <= first.time_seconds {
        return round(first.weight);
    }
    if clamped >= last.time_seconds {
        return round(last.weight);
    }
    let next_index = keyframes
        .iter()
        .position(|keyframe| keyframe.time_seconds >= clamped)
        .unwrap_or(keyframes.len() - 1);
    let next = &keyframes[next_index];
    let previous = &keyframes[next_index.saturating_sub(1)];
    let alpha = (clamped - previous.time_seconds) / (next.time_seconds - previous.time_seconds);
    round(previous.weight + (next.weight - previous.weight) * alpha)
}

fn string_field(value: &Value, key: &str) -> String {
    value
        .get(key)
        .and_then(Value::as_str)
        .unwrap_or("")
        .to_owned()
}

fn vec3_field(value: &Value, key: &str) -> [f32; 3] {
    let Some(items) = value.get(key).and_then(Value::as_array) else {
        return [0.0, 0.0, 0.0];
    };
    [
        items.first().and_then(Value::as_f64).unwrap_or(0.0) as f32,
        items.get(1).and_then(Value::as_f64).unwrap_or(0.0) as f32,
        items.get(2).and_then(Value::as_f64).unwrap_or(0.0) as f32,
    ]
}

fn round(value: f32) -> f32 {
    (value * 1_000_000.0).round() / 1_000_000.0
}
